### ISSUE TITLE
BlockingConnection uses wrong file descriptor and raises the wrong exception after rabbitmq broker stops

### ISSUE THREAD

**Issue by vitaly-krugl** [pika/pika#412]
Here is an easy way to reproduce it:
1. Start rabbitmq broker (e.g., ```rabbitmq-server -detached```)
2. Create a BlockingConnection and channel()
3. Stop rabbitmq broker (e.g., ```rabbitmqctl stop```
4. Call ```queue_declare```: ```channel.queue_declare(queue="abc", passive=True, nowait=False)```

This results in ```error: (9, 'Bad file descriptor')``` exception. It should have been ConnectionClosed exception. Same problem occurs while using BlockingChannel.consume(), and I imagine in other cases as well.

The reason that this is really, really bad is that the app is expecting the ConnectionClosed exception in such cases and also this: consider a multithreaded app, where on thread is using pika and the second thread does something else completely unrelated to pika. Let's say, the second thread may be opening/closing files, sockets, or pipes, etc. So, while processing ```queue_declare```, pika detects that the server shut down, and BlockingConnection closes the socket in ```_on_connection_closed``` by calling ```_adapter_disconnect```. However, subsequently, ```_handle_read``` is called, and it calls ```self._read_poller.ready()```. ```self._read_poller``` still has the old file descriptor, so it polls on it. Considering that file descriptor numbers get recycled (at least in unix/linux), it's quite possible to have a race condition where the second thread opens some other file or socket with that same file descriptor just before ```self._read_poller.ready()```.

**Proposed solution**
It seems like ```_adapter_disconnect()``` should also close ```self._read_poller``` (ReadPoller would need a method for this) just to be safe, and ```BlockingConnection._handle_read``` should check for self.socket being None and raise ConnectionClosed instead of calling ```self._read_poller.ready()```. I also don't understand why ```BlockingConnection._on_connection_closed()``` doesn't just go ahead and raise ConnectionClosed in this scenario.

Also, ```BlockingConnection.process_data_events``` is doing something quite hacky: it's catching AttributeError and mapping it to ConnectionClosed. What if the AttributeError is due to a bug (e.g. typo in code) and has nothing to do with the connection being closed?

**Here is the full Traceback:**

```
    577     try:
    578       r = self._channelMgr.channel.queue_declare(
--> 579         queue=mqName, passive=True, nowait=False)
    580       return r.method.message_count == 0
    581     except pika_exceptions.ChannelClosed as e:

/Users/vkruglikov/nta/nupic-darwin64/nta/eng/lib/python2.7/site-packages/pika/adapters/blocking_connection.pyc in queue_declare(self, queue, passive, durable, exclusive, auto_delete, nowait, arguments)
    865                                             exclusive, auto_delete, nowait,
    866                                             arguments or dict()),
--> 867                          None, replies)
    868 
    869     def queue_delete(self, queue='', if_unused=False, if_empty=False,

/Users/vkruglikov/nta/nupic-darwin64/nta/eng/lib/python2.7/site-packages/pika/adapters/blocking_connection.pyc in _rpc(self, method_frame, callback, acceptable_replies, content, force_data_events)
   1102         self._received_response = False
   1103         self._send_method(method_frame, content,
-> 1104                           self._wait_on_response(method_frame))
   1105         if force_data_events and self._force_data_events_override is not False:
   1106             self.connection.process_data_events()

/Users/vkruglikov/nta/nupic-darwin64/nta/eng/lib/python2.7/site-packages/pika/adapters/blocking_connection.pyc in _send_method(self, method_frame, content, wait)
   1122         while wait and not self._received_response:
   1123             try:
-> 1124                 self.connection.process_data_events()
   1125             except exceptions.AMQPConnectionError:
   1126                 break

/Users/vkruglikov/nta/nupic-darwin64/nta/eng/lib/python2.7/site-packages/pika/adapters/blocking_connection.pyc in process_data_events(self)
    213         """
    214         try:
--> 215             if self._handle_read():
    216                 self._socket_timeouts = 0
    217         except AttributeError:

/Users/vkruglikov/nta/nupic-darwin64/nta/eng/lib/python2.7/site-packages/pika/adapters/blocking_connection.pyc in _handle_read(self)
    325 
    326         """
--> 327         if self._read_poller.ready():
    328             super(BlockingConnection, self)._handle_read()
    329             self._frames_written_without_read = 0

/Users/vkruglikov/nta/nupic-darwin64/nta/eng/lib/python2.7/site-packages/pika/adapters/blocking_connection.pyc in ready(self)
     64         else:
     65             ready, unused_wri, unused_err = select.select([self.fd], [], [],
---> 66                                                           self.poll_timeout)
     67             return len(ready) > 0
     68 

error: (9, 'Bad file descriptor')
```

\
\
**And here is another Traceback during BlockingChannel.consume():**
```
--> 415             for methodFrame, _headerFrame, body in channel.consume(self._mqName):
    416               if self._stopRequested:
    417                 break

/Users/vkruglikov/nta/nupic-darwin64/nta/eng/lib/python2.7/site-packages/pika/adapters/blocking_connection.pyc in consume(self, queue)
    689             if self._generator_messages:
    690                 yield self._generator_messages.pop(0)
--> 691             self.connection.process_data_events()
    692 
    693     def force_data_events(self, enable):

/Users/vkruglikov/nta/nupic-darwin64/nta/eng/lib/python2.7/site-packages/pika/adapters/blocking_connection.pyc in process_data_events(self)
    213         """
    214         try:
--> 215             if self._handle_read():
    216                 self._socket_timeouts = 0
    217         except AttributeError:

/Users/vkruglikov/nta/nupic-darwin64/nta/eng/lib/python2.7/site-packages/pika/adapters/blocking_connection.pyc in _handle_read(self)
    325 
    326         """
--> 327         if self._read_poller.ready():
    328             super(BlockingConnection, self)._handle_read()
    329             self._frames_written_without_read = 0

/Users/vkruglikov/nta/nupic-darwin64/nta/eng/lib/python2.7/site-packages/pika/adapters/blocking_connection.pyc in ready(self)
     64         else:
     65             ready, unused_wri, unused_err = select.select([self.fd], [], [],
---> 66                                                           self.poll_timeout)
     67             return len(ready) > 0
     68 

error: (9, 'Bad file descriptor')
```

**Comment by vitaly-krugl** [comment]
Ok, with PRs #533  and #542 merged, the situation is a lot better, **but not perfect yet**.

Using the simple steps outlines at the top of this issue's description, we now correctly get `ConnectionClosed: (320, "CONNECTION_FORCED - broker forced connection closure with reason 'shutdown'")` after calling `channel.queue_declare(...)`. At this point `connection.is_open` now correctly evaluates to `False`; however, `channel.is_open` (arguably) incorrectly still evaluates to `True`.

@gmr, referring to the above, should it be a goal that if the connection is forced closed (either gracefully or due to loss of connection), then its channels should also be forced into a `channel.is_closed = True` state.

So, after the initial `ConnectionClosed` exception, if one were to attempt another operation on that channel - e.g., `channel.queue_declare(queue="abc", passive=True, nowait=False)` - then the original issue `error: (9, 'Bad file descriptor')` resurfaces as follows: 
```
channel.queue_declare(queue="abc", passive=True, nowait=False)
---------------------------------------------------------------------------
error                                     Traceback (most recent call last)
<ipython-input-10-e1c313491e58> in <module>()
----> 1 channel.queue_declare(queue="abc", passive=True, nowait=False)

/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py in queue_declare(self, queue, passive, durable, exclusive, auto_delete, nowait, arguments)
    932         return self._rpc(spec.Queue.Declare(0, queue, passive, durable,
    933                                             exclusive, auto_delete, nowait,
--> 934                                             arguments or dict()), None, replies)
    935 
    936     def queue_delete(self,

/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py in _rpc(self, method_frame, callback, acceptable_replies, content, force_data_events)
   1170             replies.append(key)
   1171         self._send_method(method_frame, content,
-> 1172                           self._wait_on_response(method_frame))
   1173         if force_data_events and self._force_data_events_override is not False:
   1174             self.connection.process_data_events()

/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py in _send_method(self, method_frame, content, wait)
   1191         while wait and not self._received_response:
   1192             try:
-> 1193                 self.connection.process_data_events()
   1194             except exceptions.ConnectionClosed:
   1195                 # No further I/O is possible, so propagate exception

/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py in process_data_events(self)
    240         """
    241         try:
--> 242             if self._handle_read():
    243                 self._socket_timeouts = 0
    244         except socket.timeout:

/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py in _handle_read(self)
    355 
    356         """
--> 357         if self._read_poller.ready():
    358             super(BlockingConnection, self)._handle_read()
    359             self._frames_written_without_read = 0

/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py in inner(*args, **kwargs)
     42         while True:
     43             try:
---> 44                 return f(*args, **kwargs)
     45             except select.error as e:
     46                 if e[0] != errno.EINTR:

/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py in ready(self)
     89         else:
     90             ready, unused_wri, unused_err = select_function([self.fd], [], [],
---> 91                                                             self.poll_timeout)
     92             return bool(ready)
     93 

error: (9, 'Bad file descriptor')
```

**Comment by vitaly-krugl** [comment]
Should be okay to close once PR #567 is merged. PR #567 resets `BlockingConnection._read_poller`, making it impossible to access the closed file descriptor

**Comment by vitaly-krugl** [comment]
Fixed via #533, #542, and #567
